pyro/ops/einsum/adjoint.py

```diff
@@ -0,0 +1,142 @@
+from __future__ import absolute_import, division, print_function
+
+import weakref
+from abc import ABCMeta, abstractmethod
+
+import torch
+from six import add_metaclass
+
+from pyro.ops import packed
+
+SAMPLE_SYMBOL = " "  # must be unique and precede alphanumeric characters
+
+
+@add_metaclass(ABCMeta)
+class Backward(object):
+    is_leaf = False
+
+    def __call__(self):
+        """
+        Performs entire backward pass in depth-first order.
+        """
+        message = None
+        stack = [(self, message)]
+        while stack:
+            bwd, message = stack.pop()
+            stack.extend(bwd.process(message))
+
+    @abstractmethod
+    def process(self, message):
+        raise NotImplementedError
+
+
+class _LeafBackward(Backward):
+    is_leaf = True
+
+    def __init__(self, target):
+        self.target = weakref.ref(target)
+
+    def process(self, message):
+        target = self.target()
+        target._pyro_backward_result = message
+        return ()
+
+
+def require_backward(tensor):
+    """
+    Marks a tensor as a leaf in the adjoint graph.
+    """
+    tensor._pyro_backward = _LeafBackward(tensor)
```

Comment:
Just for my understanding - is the backward pass repeated when we call `_pyro_backward()` twice? Would it ever be beneficial to cache it?

Reply:
Yes it should be repeated; the caching should happen at the `opt_einsum.shared_intermediates` layer. I haven't verified that this works yet, since I haven't needed to rely on it.

Reply:
Got it, thanks for explaining!